File: java/src/main/java/money_problem/domain/NewBank.java
package money_problem.domain;

import io.vavr.Function2;
import io.vavr.collection.Map;
import io.vavr.control.Either;

import static io.vavr.API.Left;
import static io.vavr.API.Right;
import static io.vavr.collection.HashMap.empty;
import static io.vavr.collection.LinkedHashMap.of;

public class NewBank {
    private final Currency pivotCurrency;
    private final Map<String, ExchangeRate> exchangeRates;

    private final Map<Function2<Money, Currency, Boolean>, Function2<Money, Currency, Money>> convert = of(
            (money, to) -> isSameCurrency(money.currency(), to), (Function2<Money, Currency, Money>) (money, to) -> money,
            this::canConvertDirectly, this::convertDirectly,
            this::canConvertThroughPivotCurrency, this::convertThroughPivotCurrency
    );

    private NewBank(Currency pivotCurrency, Map<String, ExchangeRate> exchangeRates) {
        this.pivotCurrency = pivotCurrency;
        this.exchangeRates = exchangeRates;
    }

    private NewBank(Currency pivotCurrency) {
        this(pivotCurrency, empty());
    }

    public static NewBank withPivotCurrency(Currency pivotCurrency) {
        return new NewBank(pivotCurrency);
    }

    public Either<Error, NewBank> add(ExchangeRate exchangeRate) {
        return !isSameCurrency(exchangeRate.getCurrency(), pivotCurrency)
                ? Right(addMultiplierAndDividerExchangeRate(exchangeRate))
                : Left(new Error("Can not add an exchange rate for the pivot currency"));
    }

    private boolean isSameCurrency(Currency currency, Currency otherCurrency) {
        return currency == otherCurrency;
    }

    private NewBank addMultiplierAndDividerExchangeRate(ExchangeRate exchangeRate) {
        return new NewBank(
                pivotCurrency,
                exchangeRates.put(keyFor(pivotCurrency, exchangeRate.getCurrency()), exchangeRate)
                        .put(keyFor(exchangeRate.getCurrency(), pivotCurrency), dividerRate(exchangeRate))
        );
    }

    private ExchangeRate dividerRate(ExchangeRate exchangeRate) {
        return new ExchangeRate(1 / exchangeRate.getRate(), exchangeRate.getCurrency());
    }

    private static String keyFor(Currency from, Currency to) {
        return from + "->" + to;
    }

    public Either<Error, Money> convert(Money money, Currency to) {
        return convert.find(canConvert -> canConvert._1.apply(money, to))
                .map(k -> k._2.apply(money, to))
                .toEither(new Error("No exchange rate defined for " + keyFor(money.currency(), to)));
    }

    private boolean canConvertDirectly(Money money, Currency to) {
        return exchangeRates.containsKey(keyFor(money.currency(), to));
    }

    private boolean canConvertThroughPivotCurrency(Money money, Currency to) {
        return exchangeRates.containsKey(keyFor(pivotCurrency, money.currency()))
                && exchangeRates.containsKey(keyFor(pivotCurrency, to));
    }

    private Money convertDirectly(Money money, Currency to) {
        var exchangeRate = exchangeRates.getOrElse(keyFor(money.currency(), to), new ExchangeRate(0, to));
        return new Money(money.amount() * exchangeRate.getRate(), to);
    }

    private Money convertThroughPivotCurrency(Money money, Currency to) {
        return convertDirectly(convertDirectly(money, pivotCurrency), to);
    }
}

File: c#/money-problem.Domain/NewBank.cs
using LanguageExt;
using static LanguageExt.Prelude;

namespace money_problem.Domain;

public class NewBank
{
    public const string SameExchangeRateThanCurrency = "Cannot add an exchange rate for the pivot currency.";
    private readonly Seq<NewExchangeRate> exchangeRates;
    private readonly Currency pivotCurrency;

    private NewBank(Currency pivotCurrency, Seq<NewExchangeRate> exchangeRates)
    {
        this.pivotCurrency = pivotCurrency;
        this.exchangeRates = exchangeRates;
    }

    public static NewBank WithPivotCurrency(Currency currency) => new(currency, LanguageExt.Seq<NewExchangeRate>.Empty);

    public Either<Error, NewBank> Add(NewExchangeRate exchangeRate) =>
        this.IsPivotCurrency(exchangeRate.Currency)
            ? new Error(SameExchangeRateThanCurrency)
            : this.AddExchangeRate(exchangeRate);

    private NewBank AddExchangeRate(NewExchangeRate exchangeRate) =>
        new(this.pivotCurrency, this.exchangeRates
            .Filter(element => element.Currency != exchangeRate.Currency)
            .Add(exchangeRate));

    private bool IsPivotCurrency(Currency currency) => currency == this.pivotCurrency;

    public Either<Error, Money> Convert(Money money, Currency currency) =>
        this.GetExchangeRate(money, currency)
            .Map(rate => ConvertUsingExchangeRate(money, rate))
            .Match(Either<Error, Money>.Right,
                () => new Error(FormatMissingExchangeRate(money.Currency, currency)));

    private Option<NewExchangeRate> GetExchangeRate(Money money, Currency currency)
    {
        if (money.HasCurrency(currency))
        {
            return GetExchangeRateForSameCurrency(currency);
        }

        if (this.IsPivotCurrency(currency))
        {
            return this.FindExchangeRate(money.Currency).Bind(this.ReverseExchangeRate);
        }

        var exchangeSource = this.FindExchangeRate(money.Currency);
        var exchangeDestination = this.FindExchangeRate(currency);
        return this.ShouldConvertThroughPivotCurrency(money.Currency, currency)
            ? ComputeExchangeRate(exchangeSource, exchangeDestination)
            : exchangeDestination;
    }

    private Option<NewExchangeRate> ReverseExchangeRate(NewExchangeRate exchange) =>
        NewExchangeRate
            .From(this.pivotCurrency, exchange.GetReversedRate())
            .Match(Some, _ => Option<NewExchangeRate>.None);

    private static Option<NewExchangeRate> GetExchangeRateForSameCurrency(Currency currency) =>
        NewExchangeRate
            .From(currency, 1)
            .Match(Some, _ => Option<NewExchangeRate>.None);

    private bool ShouldConvertThroughPivotCurrency(Currency source, Currency destination) =>
        !this.IsPivotCurrency(source) && !this.IsPivotCurrency(destination);

    private static Option<NewExchangeRate> ComputeExchangeRate(Option<NewExchangeRate> source,
        Option<NewExchangeRate> destination) =>
        destination
            .Bind(exchange => source
                .Map(sourceExchange => sourceExchange.GetReversedRate())
                .Map(rate => rate * exchange.Rate)
                .Map(rate => new Tuple<Currency, double>(exchange.Currency, rate)))
            .Map(exchange => NewExchangeRate
                .From(exchange.Item1, exchange.Item2)
                .IfLeft(_ => NewExchangeRate.Default(exchange.Item1)));

    private static Money ConvertUsingExchangeRate(Money money, NewExchangeRate exchangeRate) =>
        new(money.Amount * exchangeRate.Rate, exchangeRate.Currency);

    private Option<NewExchangeRate> FindExchangeRate(Currency currency) =>
        this.exchangeRates.Find(exchange => exchange.Currency == currency);

    private static string FormatMissingExchangeRate(Currency from, Currency to) => $"{from}->{to}.";
}